src/infra/auth-rebac/inmem/src/repos/rebac_repository_inmem.rs

```diff
@@ -0,0 +1,495 @@
+
+        let maybe_row_ids = readable_state
+            .index_subject_entity_object_type
+            .get(&index_hash);
+
+        let Some(row_ids) = maybe_row_ids else {
+            return Err(SubjectEntityRelationsByObjectTypeError::not_found(
+                subject_entity,
+                object_entity_type,
+            ));
+        };
+
+        let rows = self.get_entities_relations_by_row_ids(row_ids, &readable_state);
+        let res = rows
+            .into_iter()
+            .map(|row| ObjectEntityWithRelation {
+                entity_type: row.object_entity.entity_type,
+                entity_id: row.object_entity.entity_id.clone().into(),
+                relation: row.relationship,
+            })
+            .collect();
+
+        Ok(res)
+    }
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetRelationsBetweenEntitiesError> {
+        let index_hash =
+            EntityHasher::subject_entity_object_entity_index_hash(subject_entity, object_entity);
+
+        let readable_state = self.entities_relations_state.read().await;
+
+        let maybe_row_ids = readable_state
+            .index_subject_entity_object_entity
+            .get(&index_hash);
+
+        let Some(row_ids) = maybe_row_ids else {
+            return Err(GetRelationsBetweenEntitiesError::not_found(
+                subject_entity,
+                object_entity,
+            ));
+        };
+
+        let rows = self.get_entities_relations_by_row_ids(row_ids, &readable_state);
+        let res = rows.into_iter().map(|row| row.relationship).collect();
+
+        Ok(res)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+struct EntityRow {
+    pub entity_type: EntityType,
+    pub entity_id: String,
+}
+
+impl<'a> From<Entity<'a>> for EntityRow {
+    fn from(value: Entity<'a>) -> Self {
+        Self {
+            entity_type: value.entity_type,
+            entity_id: value.entity_id.to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct EntitiesRelationsRow {
+    #[allow(dead_code)]
+    subject_entity: EntityRow,
+    relationship: Relation,
+    object_entity: EntityRow,
+}
+
+impl EntitiesRelationsRow {
+    pub fn new(
+        subject_entity: Entity<'_>,
+        relationship: Relation,
+        object_entity: Entity<'_>,
+    ) -> Self {
+        Self {
+            subject_entity: subject_entity.into(),
+            relationship,
+            object_entity: object_entity.into(),
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// EntityHasher
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+struct EntityHasher {}
+
+impl EntityHasher {
```

Comment:
This is surprisingly complex...

If I'm not mistaking, unless this is a perfect hash - you have a possibility of collisions that go undetected.

Why not simply use `HashMap<Entity<'static>, ...>`?

Honestly, even storing the whole table as `Vec<(Entity, Relation, Entity)>` with linear search complexity would be fine for the first in-mem implementation.

Reply:
I might be wrong, but my impression is that using `Cow` in `Entity` type costed you a whole bunch of time that could be better spent elsewhere e.g. derisking OSO integration. 

I like idiomatic Rust, but this kind of optimizations can always be done later when there's a need.

Reply:
> If I'm not mistaking, unless this is a perfect hash - you have a possibility of collisions that go undetected.

I used the default hasher used in `HashSet` (which is based on `HashMap`)

> Why not simply use `HashMap<Entity<'static>, ...>`?

I wanted to get away from the cloning that occurs even at times when lookups -- but, yeah, it was hard to read...

> Honestly, even storing the whole table as Vec<(Entity, Relation, Entity)> with linear search complexity would be fine for the first in-mem implementation.

It will also be linear time for the search, but used `HashSet<T>`. And in general, simplified the implementation, without programmatically created indexes.

> I might be wrong, but my impression is that using `Cow` in `Entity` type costed you a whole bunch

No, `Cow<T>` has nothing to do with it -- I rather wanted to get away from O(N), as well as storing duplicate records in indexes. This implementation didn't take much time.

Still, it's hard to deny that it turned out to be more difficult for the reader than I wanted it to be.

> I like idiomatic Rust, but this kind of optimizations can always be done later when there's a need.

I tried to keep the interfaces also ideomatic, but with simpler implementation

-- 

Fixed

Reply:
> I used the default hasher used in HashSet (which is based on HashMap)

When you do a lookup in a hash map it first looks up a value by hash and then does `Eq` test to check that the right key was found and not just a key with the same hash.

My point was that by using pre-hashed values instead of `Entity<'static>` as keys the `Eq` test would no longer guarantee correct functionality under collisions.

Just making sure we understood one another.